synapse/storage/databases/main/monthly_active_users.py

```diff
@@ -122,6 +122,49 @@ def _count_users_by_service(txn: LoggingTransaction) -> Dict[str, int]:
             "count_users_by_service", _count_users_by_service
         )
 
+    async def get_monthly_active_users_by_service(
+        self, start_timestamp: Optional[int] = None, end_timestamp: Optional[int] = None
+    ) -> List[Tuple[str, str]]:
+        """Generates list of monthly active users and their services.
+        Please see "get_monthly_active_count_by_service" docstring for more details
+        about services.
+
+        Arguments:
+            start_timestamp: If specified, only include users that were first active
+                at or after this point
+            end_timestamp: If specified, only include users that were first active
+                at or before this point
+
+        Returns:
+            A list of tuples (appservice_id, user_id)
+
+        """
+        if start_timestamp is not None and end_timestamp is not None:
+            where_clause = 'WHERE "timestamp" >= ? and "timestamp" <= ?'
+            query_params = [start_timestamp, end_timestamp]
+        elif start_timestamp is not None:
+            where_clause = 'WHERE "timestamp" >= ?'
+            query_params = [start_timestamp]
+        elif end_timestamp is not None:
+            where_clause = 'WHERE "timestamp" <= ?'
+            query_params = [end_timestamp]
+        else:
+            where_clause = ""
+            query_params = []
+
+        def _list_users(txn: LoggingTransaction) -> List[Tuple[str, str]]:
+            sql = f"""
+                    SELECT COALESCE(appservice_id, 'native'), user_id
```

Comment:
Can we just return `null` if there isn't an appservice? It saves having to use a magic string

Reply:
We could live with that, but the pre-existing [get_monthly_active_count_by_service](https://github.com/matrix-org/synapse/blob/28199e93579b5a73841a95ed4d355322227432b5/synapse/storage/databases/main/monthly_active_users.py#L111) method (which this new method is modelled on) is emitting `"native"`.

Reply:
Oh interesting, fair enough then. I think it probably does make more sense to be consistent then